### PR TITLE
adding Coderabbit

### PR DESCRIPTION
**Pull request by BradReesWork**
added the config files for running Coderabbit

### CODE REVIEW

File: cpp/REVIEW_GUIDELINES.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+- Kernel launch with zero blocks/threads or invalid grid/block dimensions
+
+### Algorithm Correctness
+- Logic errors producing wrong results
+- Off-by-one errors in index arithmetic or kernel launch bounds
+- Incorrect null handling for fixed-width columns (null values are undefined, must not be read)
+- Accessing offsets child of an empty string or list column
+- Nested type columns (LIST, STRUCT) not sanitized (libcugraph expects sanitized null masks)
+- Incorrect use of `cugraph::size_type` (signed 32-bit) for sizes
+- Offsets must be `int32_t` or `int64_t` as appropriate; only `int32_t` for LIST offsets
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: These look like `cudf` requirements. Does not apply to `cugraph`.

File: cpp/REVIEW_GUIDELINES.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+### Performance Issues
+- Unnecessary host-device synchronization blocking GPU pipeline
+- Missing `rmm::exec_policy_nosync(stream)` for Thrust device execution
+- Multiple levels of `type_dispatcher` (avoid when possible)
+- Raw loops or raw kernels where CUB/Thrust/STL algorithms suffice
+- Suboptimal memory access patterns (non-coalesced, strided, unaligned)
+- Excessive memory allocations in hot paths
+- Warp divergence in compute-heavy kernels
+- Shared memory bank conflicts
+- Host-side compute-intensive algorithms not using `cugraph::detail::host_worker_pool()`
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: Sounds like a `cudf` specific requirement. `cugraph` doesn't have `host_worker_pool()`.

File: cpp/REVIEW_GUIDELINES.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+### Memory Management Violations
+- Returned memory not using the passed-in memory resource (MR)
+- Temporary memory not using `cugraph::get_current_device_resource_ref()`
+- Using `thrust::device_vector` or `rmm::device_scalar<T>` instead of `rmm::device_uvector` or `cugraph::detail::device_scalar<T>`
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: This requires some discussion. So, we are not expected to use `rmm::device_scalar`? (`cugraph` doesn't have `detail::device_scalar`, this sounds like a cudf specific thing).

File: cpp/REVIEW_GUIDELINES.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+- Using raw `cudaMemcpyAsync` instead of `cugraph::detail::cuda_memcpy_async` / `memcpy_async` / `memcpy_batch_async`
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md:55]: This also sounds like a cudf thing. We use thrust::copy or raft::copy.

File: cpp/REVIEW_GUIDELINES.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+- Using `cugraph::detail::make_host_vector{,_async}` instead of `cugraph::detail::make_pinned_vector{,_async}` for small H2D/D2H transfers
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: This also sounds like  a cudf specific requirement.

File: cpp/REVIEW_GUIDELINES.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+- Async memcpy staging buffers that don't outlive the copy
+
+### API & Design
+- Stream and MR not as last two parameters (stream before MR); public APIs must have defaults, detail APIs must not
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md:59]: We often use handle instead of stream and mr. This also sounds like a cudf thing.

File: cpp/REVIEW_GUIDELINES.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+- Missing `CUGRAPH_FUNC_RANGE()` in public functions before delegating to `detail::`
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md:60]: Sounds like a cudf requirement.

File: cpp/REVIEW_GUIDELINES.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+- Missing `[[nodiscard]]` on side-effect-free functions returning non-void results
+
+### Concurrency & Stream Safety
+- Stream and MR parameters not propagated across all internal APIs
+- Implicit CUDA default stream use outside tests, benchmarks, and public API default parameters
+- Unnecessary `cudaDeviceSynchronize()`; use `stream.synchronize()` when synchronization is required
+- Operations incorrectly ordered on different streams without events or explicit dependencies
+
+### Type Dispatch Patterns
+- New dispatch functors using `CUGRAPH_ENABLE_IF` instead of C++20 `requires` clauses
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: We don't have `CUGRAPH_ENABLE_IF`. `CUDF_ENABLE_IF` is a cudf specific thing.

File: cpp/REVIEW_GUIDELINES.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+- Unsupported type overloads not calling `CUGRAPH_FAIL` or `CUGRAPH_UNREACHABLE`
+- Missing `static constexpr bool is_supported()` helper where useful
+
+### Test Quality
+- Tests missing edge cases: empty input, null values, sliced columns, boundary sizes, multi-block sizes
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: `sliced columns`, `multi-block sizes` sound cudf specific.

File: cpp/REVIEW_GUIDELINES.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+- String tests missing non-ASCII UTF-8 characters
+- Decimal types in `FixedWidthTypes` but not `NumericTypes` — verify correct type list usage
+- Tests not using `#include <cugraph_test/cugraph_gtest.hpp>` (never raw `gtest/gtest.h`)
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: Sounds like a cudf specific requirement.

File: cpp/REVIEW_GUIDELINES.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+- Test code not in the global namespace
+- Benchmarks not using NVBench (not Google Benchmark)
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: AFAIK, we are currently using Google benchmark.

File: cpp/REVIEW_GUIDELINES.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+**CRITICAL** (unchecked CUDA error):
+```
+CRITICAL: Unchecked kernel launch
+
+Issue: Kernel launch error not checked
+Why: Subsequent operations assume success, causing silent corruption
+
+Suggested fix:
+myKernel<<<grid, block, 0, stream.value()>>>(args);
+CUGRAPH_CUDA_TRY(cudaGetLastError());
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: CUGRAPH_CUDA_TRY should be RAFT_CUDA_TRY in our context.

File: cpp/REVIEW_GUIDELINES.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+```
+
+**CRITICAL** (null handling):
+```
+CRITICAL: Reading null values of fixed-width column
+
+Issue: Accessing element values without checking null mask
+Why: Null values of fixed-width columns are undefined; reading them is undefined behavior
+```
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: Sounds like a cudf thing.

File: cpp/REVIEW_GUIDELINES.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+- "This function could be split into smaller functions"
+
+---
+
+## C++/CUDA-Specific Considerations
+
+**Error Handling**:
+- Use cugraph macros: `CUGRAPH_EXPECTS`, `CUGRAPH_FAIL`, `CUGRAPH_UNREACHABLE`
+- `CUGRAPH_EXPECTS` condition must be a pure predicate with no side effects
+- Use `CUGRAPH_CUDA_TRY` for CUDA API calls; `CUGRAPH_CUDA_TRY_NO_THROW` in destructors
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: `CUGRAPH_CUDA_TRY` and `CUGRAPH_CUDA_TRY_NO_THROW`, `CUGRAPH`=>`RAFT`.

File: cpp/REVIEW_GUIDELINES.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+- Every CUDA call must have error checking (kernel launches, memory ops, sync)
+
+**Memory Management**:
+- Use `rmm::device_uvector` for typed device memory (not `thrust::device_vector`)
+- Use `cugraph::detail::device_scalar<T>` (not `rmm::device_scalar<T>`)
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: We are directly using rmm::device_scalar.

File: cpp/REVIEW_GUIDELINES.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+- Returned memory uses the passed-in MR; temporary memory uses `cugraph::get_current_device_resource_ref()`
+- Use `cugraph::detail::cuda_memcpy_async` / `memcpy_async` / `memcpy_batch_async` (not raw `cudaMemcpyAsync`)
+- Prefer `cugraph::detail::make_pinned_vector{,_async}` over `make_host_vector{,_async}` for small H2D/D2H transfers
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: cugraph doesn't have get_current_device_resource_ref.

Same for cuda_memcpy_async, memcpy_async, memcpy_batch_async, make_pinned_vector, make_host_vector.

File: cpp/REVIEW_GUIDELINES.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+**Stream Management**:
+- Stream and MR as last two parameters (stream before MR)
+- Public APIs have default values; detail APIs do not
+- Disallow implicit default stream use except in tests, benchmarks, and public API default parameters
+- All kernel launches and Thrust calls must use the stream parameter
+- Use `rmm::exec_policy_nosync(stream)` for all Thrust device execution
+
+**CUDA Kernels**:
+- Use `CUGRAPH_KERNEL` macro (not raw `__global__`), preferably with `__launch_bounds__`
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: We don't have CUGRAPH_KERNEL

File: cpp/REVIEW_GUIDELINES.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+- Use `cuda::std::` types/algorithms in device code (`cuda::std::min`, `cuda::std::pair`, etc.)
+- Use `cuda::make_constant_iterator` over `thrust::make_constant_iterator` for device-side constant iterators
+- Use `cuda::proclaim_return_type<T>(lambda)` when passing device lambdas to `make_counting_transform_iterator`
+- Prefer modern CUDA C++ primitives: `cuda::std::popcount` over `__popc`, `cg::thread_block::thread_rank()` over `threadIdx.x`
+
+**Public API** (`cpp/include/cugraph/`, `cpp/include/nvtext/`):
+- Functions must have `cugraph_EXPORT`
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: `cugraph_EXPORT`=>`CUGRAPH_EXPORT`

File: cpp/REVIEW_GUIDELINES.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+- Doxygen documentation required (`@brief`, `@param`, `@return`, `@throw`, `@tparam`)
+- Public functions: `CUGRAPH_FUNC_RANGE()` then delegate to `detail::` (trivial functions may skip)
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: We don't have CUGRAPH_FUNC_RANGE.

File: cpp/REVIEW_GUIDELINES.md [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+- API changes require deprecation warnings (`[[deprecated]]`, `@deprecated`, PR labels)
+- Use `[[nodiscard]]` on side-effect-free functions with non-void return
+
+**C++ Style**:
+- Prefer C++20 `requires` clauses over `CUGRAPH_ENABLE_IF` for type-gating dispatch functors
```
Comment (seunghwak) [inline cpp/REVIEW_GUIDELINES.md]: We don't have CUGRAPH_ENABLE_IF.

**Review by seunghwak — COMMENTED** [review]
It seems like the C++ requirements include many cudf specific items. We need to delete them. Otherwise, looks like a good starting point.

File: cpp/REVIEW_GUIDELINES.md (hunk 5 above)
Comment (seunghwak) [reply cpp/REVIEW_GUIDELINES.md]: We don't have make_host_vector or make_pinned_vector.

**Review by seunghwak — APPROVED** [review]
We don't have 

`cugraph::detail::make_host_vector{,_async}` or `cugraph::detail::make_pinned_vector{,_async}`, so the line mentioning these functions is irrelevant.

Otherwise, LGTM.

File: cpp/REVIEW_GUIDELINES.md [hunk 21]
```diff
@@ -0,0 +1,226 @@
+# AI Code Review Guidelines - cuGraph C++/CUDA
+
+**Role**: Act as a principal engineer with 10+ years experience in GPU computing, Modern C++, and high-performance data processing. Strongly prefer modern C++ and established CUDA/C++ library algorithms over raw loops. Focus ONLY on CRITICAL and HIGH issues.
+
+**Target**: Sub-3% false positive rate. Be direct, concise, minimal.
+
+**Context**: cuGraph C++ layer (libcugraph) provides GPU-accelerated Graph operations using CUDA, with dependencies on RMM, thrust, and CUB.
```
Comment (ChuckHastings) [inline cpp/REVIEW_GUIDELINES.md]: Might want to add raft and cuvs as dependencies here.

File: cpp/REVIEW_GUIDELINES.md [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+
+### Resource Management
+- GPU memory leaks (device allocations, managed memory, pinned memory)
+- CUDA stream/event leaks or improper cleanup
+- Missing RAII or proper cleanup, including in exception paths
+- Raw owning pointers instead of `std::unique_ptr`, `std::shared_ptr`, `std::reference_wrapper`
+
+### API Breaking Changes
+- C++ API changes without proper deprecation warnings
+- Changes to data structures exposed in public headers (`cpp/include/cugraph/`, `cpp/include/nvtext/`)
```
Comment (ChuckHastings) [inline cpp/REVIEW_GUIDELINES.md]: remove nvtext, add cugraph_c instead

File: cpp/REVIEW_GUIDELINES.md [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+- Missing `[[deprecated]]` attribute, `@deprecated` doxygen tag, or PR labels ("deprecation" / "breaking")
+
+## HIGH Issues (Comment if Substantial)
+
+### Performance Issues
+- Unnecessary host-device synchronization blocking GPU pipeline
+- Missing `rmm::exec_policy_nosync(stream)` for Thrust device execution
```
Comment (ChuckHastings) [inline cpp/REVIEW_GUIDELINES.md]: add "or handle.get_thrust_policy()" as an option here.

File: cpp/REVIEW_GUIDELINES.md [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+- Multiple levels of `type_dispatcher` (avoid when possible)
```
Comment (ChuckHastings) [inline cpp/REVIEW_GUIDELINES.md]: This is a cudf only concern, we should drop this.

File: cpp/REVIEW_GUIDELINES.md [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+- Missing `CUGRAPH_FUNC_RANGE()` in public functions before delegating to `detail::`
+- Missing `CUGRAPH_EXPORT` on public API functions in `cpp/include/cugraph/`
+- `CUGRAPH_EXPECTS` condition with side effects (must be a pure predicate)
+- Functions defined in headers that are neither templated nor `inline` device functions
+- Anonymous namespaces in headers (must only be in single-TU `.cpp`/`.cu` files)
+- Owning vectors passed by copy/reference instead of being moved (transferring ownership)
+- Missing `[[nodiscard]]` on side-effect-free functions returning non-void results
+
+### Concurrency & Stream Safety
+- Stream and MR parameters not propagated across all internal APIs
```
Comment (ChuckHastings) [inline cpp/REVIEW_GUIDELINES.md:69]: Using handle should be reasonable here as well.

File: cpp/REVIEW_GUIDELINES.md (hunk 5 above)
Comment (BradReesWork) [reply cpp/REVIEW_GUIDELINES.md]: removed

File: cpp/REVIEW_GUIDELINES.md [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+- Implicit CUDA default stream use outside tests, benchmarks, and public API default parameters
+- Unnecessary `cudaDeviceSynchronize()`; use `stream.synchronize()` when synchronization is required
+- Operations incorrectly ordered on different streams without events or explicit dependencies
+
+### Type Dispatch Patterns
```
Comment (ChuckHastings) [inline cpp/REVIEW_GUIDELINES.md:72]: I'd leave this section out.  We have a totally different type dispatch pattern from cudf.  I'm inclined to add those tests later.

File: cpp/REVIEW_GUIDELINES.md (hunk 22 above)
Comment (BradReesWork) [reply cpp/REVIEW_GUIDELINES.md]: good catch, thanks
